### PR TITLE
164 extracting subsets of data from large external sources

### LINKED ISSUE #164: Support for extracting subsets of data from large external sources
See also #140, #154, #156.

This issue is to provide an interface which allows clients to access subsets of data from large series stored outside of a database.

A typical scenario is allowing access to only certain sections of certain datasets which are constituents of a remotely-stored HDF file.  In this case, it is assumed that the file is sufficiently large that retrieving the entire of it is impractical.

This facility should expose an interface to Hydra clients which allows:

1. Access to subsets of remote datasets
2. A 'paging' mechanism which describes the location within a dataset of an extracted portion
3. Transparent support for different data storage formats.  This should be implemented in a modular way which allows Hydra to use a common interface for accessing a range of different storage backends.
4. (As required) Some form of metadata caching which avoids repeated discovery of the structure of remote files.

This system should integrate with or extend that implemented in #154, while preserving the interface this presents to Hydra code.

### PR DESCRIPTION
**Pull request by pmslavin**
Closes #164

* Adds api to query HDf files in local and remote storage, describe datasets, and retrieve segments of datasets.
* Allows rewriting of urls to a local filestore
* Adds `hydra.ini` config to disable HDF support
* Adds some utility functions to test existence of files at urls and total filesize
* Adds tests

### CODE REVIEW

File: hydra_base/lib/data.py [hunk 1]
```diff
@@ -1092,3 +1095,53 @@ def delete_dataset(dataset_id,**kwargs):
...
+
+def get_hdf_dataframe(url, dataset_name, start, end, **kwargs):
+    """
+      Returns the rows from <start> to <end> of <dataset_name> in
+      the hdf file <url> as the json representation of a Pandas
+      DataFrame. This may then be read directly in a client with
+      pandas.read_json().
+      Raises ValueError on bad url, dataset name or bounds
+    """
+    hdf = HdfStorageAdapter()
```
Comment (knoxsp) [inline hydra_base/lib/data.py:1128]: Should there be user ID validation here to check the user can access the data?

Comment (pmslavin) [reply hydra_base/lib/data.py:1128]: There should definitely be validation that the user is permitted to access the HDF file *somewhere* in the chain between a client requesting a Hydra dataset and the contents of an HDF file being returned.

At the moment, there is the usual validation when the Hydra dataset which contains a reference to an HDF file is loaded, so if the user can't see the Hydra dataset, they also can't see the HDF url.

There is currently the potential for a user to edit the HDF url in a Hydra dataset which they have access to so that it points to any arbitrary HDF file.  Because they can see the Hydra dataset, they can then see the HDF file.

A minimal improvement here is checking that the HDF url is within an "authorised path", e.g. it goes to "s3://hydra/... rather than "s3://arbitrary-third-party/..." but then we need to pre-register paths and/or ensure that all HDF files are only ever in "authorised" places.  An easy but restrictive way to do this would be to extend the current "filestore" mechanism to require that any HDF file must be in it, ie. no external internet access.

Or particular users can be "HDF enabled" and only they have access to the HDF api.

The other end of the spectrum is per-user per-HDF access control, i.e. keeping a record of which HDF files are accessible to each user *in addition* to which Hydra datasets are accessible to each user.  The user may be able to see the Hydra dataset, but necessarily see the HDF file it references.  This solves the issue of a user rewriting the HDF url in a Hydra dataset which they can access, but is quite a heavyhanded approach.

So in summary, yes there are a few ways access to HDF files can be tightened rather than just having it piggyback on Hydra dataset validation as at present, but there are policy and admin implications for each.
